core/editor/src/dispatcher/mod.rs

```diff
@@ -55,32 +55,39 @@ impl Dispatcher {
 				}));
 
 				self.emit_response(Response::UpdateCanvas { document: state.document.render() });
+				state.tool_state.active_tool().handle_input(event);
 				Ok(())
 			}
 			Event::MouseMovement(pos) => {
-				state.tools.mouse_state.position = pos;
-				state.tools.record_trace_point();
+				state.tool_state.mouse_state.position = pos;
+				// The state has changed so we add a trace point.
+				state.tool_state.record_trace_point();
```

Comment:
You can delete every reference to Trace that is currently in the `ToolState`.
The `Trace` will be handled by the different tool.
I just wanted you to keep the modifications to the mod_keys/mouse_keys to avoid weird issues with holding a modifier key while changing the tool.

Reply:
If you hold down the shift key, then switch to a different tool, release the shift key and switch back to the first tool, the first tool never got the notification that the key was released